src/modelfree/shaping_wrappers.py

```diff
@@ -0,0 +1,208 @@
+from collections import defaultdict
+import json
+
+import numpy as np
+from stable_baselines import logger
+from stable_baselines.common.base_class import BaseRLModel
+from stable_baselines.common.vec_env import VecEnvWrapper
+
+from modelfree import envs  # noqa - needed to register sumo_auto_contact
+from modelfree.scheduling import LinearAnnealer
+
+
+class RewardShapingEnv(VecEnvWrapper):
+    """A more direct interface for shaping the reward of the attacking agent."""
+
+    default_shaping_params = {
+        'reward_center': 1,
+        'reward_ctrl': -1,
+        'reward_contact': -1,
+        'reward_survive': 1,
+
+        # sparse reward field as per gym_compete/new_envs/multi_agent_env:151
+        'reward_remaining': 1,
+        # dense reward field as per gym_compete/new_envs/agents/humanoid_fighter:45
+        # 'reward_move': 10,
+    }
+
+    def __init__(self, env, shaping_params=default_shaping_params,
+                 reward_annealer=None):
+        super().__init__(env)
+        self.shaping_params = shaping_params
+        self.reward_annealer = reward_annealer
+        self.counter = 0
+
+        self.ep_rew_dict = defaultdict(list)
+        self.step_rew_dict = defaultdict(lambda: [[] for _ in range(self.num_envs)])
+
+    def _log_sparse_dense_rewards(self):
+        if self.counter == 2048 * self.num_envs:
```

Comment:
The semantics of logging every `2048 * self.num_envs` timesteps seem odd to me. Main issues are:

  * No guarantee that this lines up with episode boundaries, so we can have variable number of episodes in our mean.
  * This number of timesteps does not correspond to anything else in the algorithm, so it's another number the user has to remember. If we wanted to log after a fixed number of timesteps, the batch size would be a natural metric.
  * There's no way to change the logging period.
  * What is `2048` anyway (avoid magic numbers).